src/pipecat/services/aws_nova_sonic/aws.py

```diff
@@ -0,0 +1,973 @@
+        await super().stop(frame)
+        await self._disconnect()
+
+    async def cancel(self, frame: CancelFrame):
+        await super().cancel(frame)
+        await self._disconnect()
+
+    #
+    # conversation resetting
+    #
+
+    async def reset_conversation(self):
+        logger.debug("Resetting conversation")
+        await self._disconnect()
+        await self._start_connecting()
+        # Use existing context
+        self._context_available = True
+        await self._finish_connecting_if_context_available()
+
+    #
+    # frame processing
+    #
+
+    async def process_frame(self, frame: Frame, direction: FrameDirection):
+        await super().process_frame(frame, direction)
+
+        if isinstance(frame, OpenAILLMContextFrame):
+            await self._handle_context(frame.context)
+        elif isinstance(frame, InputAudioRawFrame):
+            await self._handle_input_audio_frame(frame)
+        elif isinstance(frame, BotStoppedSpeakingFrame):
+            await self._handle_bot_stopped_speaking()
+        elif isinstance(frame, AWSNovaSonicFunctionCallResultFrame):
+            await self._handle_function_call_result(frame)
+        # TODO: do we need to do anything for the below four frame types?
+        elif isinstance(frame, StartInterruptionFrame):
+            # print("[pk] StartInterruptionFrame")
+            pass
+        elif isinstance(frame, UserStartedSpeakingFrame):
+            # print("[pk] UserStartedSpeakingFrame")
+            pass
+        elif isinstance(frame, StopInterruptionFrame):
+            # print("[pk] StopInterruptionFrame")
+            pass
+        elif isinstance(frame, UserStoppedSpeakingFrame):
+            # print("[pk] UserStoppedSpeakingFrame")
+            pass
+
+        await self.push_frame(frame, direction)
+
+    async def _handle_context(self, context: OpenAILLMContext):
+        # TODO: reset connection if needed (if entirely new context object provided, for instance)
+        print(f"[pk] received updated context: {context.get_messages_for_initializing_history()}")
+        if not self._context:
+            # We got our initial context - try to finish connecting
+            self._context = AWSNovaSonicLLMContext.upgrade_to_nova_sonic(
+                context, self._system_instruction
+            )
+            self._context_available = True
+            await self._finish_connecting_if_context_available()
+
+    async def _handle_input_audio_frame(self, frame: InputAudioRawFrame):
+        # Wait until we're done sending the assistant response trigger audio before sending audio
+        # from the user's mic
+        if self._triggering_assistant_response:
+            return
+
+        # TODO: check if _audio_input_paused? what causes that?
+        await self._send_user_audio_event(frame.audio)
+
+    async def _handle_bot_stopped_speaking(self):
+        # Protect against back-to-back BotStoppedSpeaking calls, which I've observed
```

Comment:
Interesting, I wouldn't expect to receive the `BotStoppedSpeakingFrame` multiple times. Is it easy to reproduce ?